lib/helpers.js

```diff
@@ -97,36 +100,35 @@ function sanitizeDocument(schema, options, doc) {
 
   // Check to see if we're going to be inserting the permissions info
   if (options.permissions) {
-    embedPermissions(schema, options, doc);
+    await embedPermissions(schema, options, doc);
   }
 
   // Apply the rules down one level if there are any path specific permissions
-  _.each(schema.pathsWithPermissionedSchemas, (path, subSchema) => {
+  await Promise.all(_.map(schema.pathsWithPermissionedSchemas, async (path, subSchema) => {
     if (innerDoc[path]) {
       // eslint-disable-next-line no-use-before-define
-      innerDoc[path] = sanitizeDocumentList(subSchema, options, innerDoc[path]);
+      innerDoc[path] = await sanitizeDocumentList(subSchema, options, innerDoc[path]);
     }
-  });
+  }));
 
   return doc;
 }
 
-function sanitizeDocumentList(schema, options, docs) {
+async function sanitizeDocumentList(schema, options, docs) {
   const multi = _.isArrayLike(docs);
   const docList = _.castArray(docs);
 
-  const filteredResult = _.chain(docList)
-    .map((doc) => {
+  const filteredResult = _.chain(await Promise.all(_.map(docList, async (doc) => {
```

Comment:
You can break this up into separate statements so it's easier to read and less nested.

Reply:
I was just trying to change as little as possible, I don't usually use underscore/lodash. Happy to break it up though.